### PR TITLE
[query] Add config for timeseries limit returned by single DB node

### PR DESCRIPTION
**Pull request by robskillington**
**What this PR does / why we need it**:

Adds a config and a sane default for setting a max timeseries limit returned by a single DB node to m3query.

**Special notes for your reviewer**:

**Does this PR introduce a user-facing and/or backwards incompatible change?**:

```release-note
NONE
```

**Does this PR require updating code package or user-facing documentation?**:

```documentation-note
NONE
```

### CODE REVIEW

File: src/cmd/services/m3query/config/config.go [hunk 1]
```diff
@@ -217,15 +228,33 @@ type PerQueryLimitsConfiguration struct {
...
 }
 
+// AsFetchOptionsBuilderOptions converts this configuration to
+// handler.FetchOptionsBuilderOptions.
+func (l *PerQueryLimitsConfiguration) AsFetchOptionsBuilderOptions() handler.FetchOptionsBuilderOptions {
+	if l.MaxTimeseries <= 0 {
+		return handler.FetchOptionsBuilderOptions{
+			Limit: defaultStorageQueryLimit,
+		}
+	}
```
Comment (benraskin92) [inline src/cmd/services/m3query/config/config.go]: nit: newline

File: src/cmd/services/m3query/config/config.go [hunk 2]
```diff
@@ -217,15 +228,33 @@ type PerQueryLimitsConfiguration struct {
 	// this field directly.
 	PrivateMaxComputedDatapoints int64 `yaml:"maxComputedDatapoints"`
 
-	// MaxFetchedDatapoints limits the number of datapoints actually used by a given query.
+	// MaxFetchedDatapoints limits the number of datapoints actually used by a
+	// given query.
 	MaxFetchedDatapoints int64 `yaml:"maxFetchedDatapoints"`
+
+	// MaxTimeseries limits the number of time series returned by a storage node.
+	MaxTimeseries int64 `yaml:"maxTimeseries"`
```
Comment (benraskin92) [inline src/cmd/services/m3query/config/config.go]: `maxFetchedTimeseries` instead to be consistent with `maxFetchedDatapoints`?

File: src/query/api/v1/handler/fetch_options.go [hunk 3]
```diff
@@ -0,0 +1,67 @@
...
+) (*storage.FetchOptions, *xhttp.ParseError) {
+	fetchOpts := storage.NewFetchOptions()
+	fetchOpts.Limit = b.opts.Limit
+	if str := req.Header.Get(LimitMaxTimeseriesHeader); str != "" {
+		n, err := strconv.Atoi(str)
+		if err != nil {
+			return nil, xhttp.NewParseError(err, http.StatusBadRequest)
+		}
+		fetchOpts.Limit = n
+	}
```
Comment (benraskin92) [inline src/query/api/v1/handler/fetch_options.go]: nit: newline

File: src/query/api/v1/handler/fetch_options.go [hunk 4]
```diff
@@ -0,0 +1,67 @@
...
+) FetchOptionsBuilder {
+	return fetchOptionsBuilder{opts: opts}
+}
+
+func (b fetchOptionsBuilder) NewFetchOptions(
+	req *http.Request,
+) (*storage.FetchOptions, *xhttp.ParseError) {
+	fetchOpts := storage.NewFetchOptions()
+	fetchOpts.Limit = b.opts.Limit
+	if str := req.Header.Get(LimitMaxTimeseriesHeader); str != "" {
```
Comment (benraskin92) [inline src/query/api/v1/handler/fetch_options.go]: Good call on making this part of the request. Although does it make more sense to have this be part of the url parameters? More of an educational question for me..

Comment (richardartoul) [inline src/query/api/v1/handler/fetch_options.go]: Do we have any documentation right now on setting query limits and such? If not would be nice to add a simple document that outlines the YAML configs you can set as well as this head.

File: src/query/api/v1/handler/headers.go [hunk 5]
```diff
@@ -21,18 +21,22 @@
...
+	// ServedByHeader is the M3 query storage execution breakdown.
 	ServedByHeader = "M3-Storage-By"
 
-	// DeprecatedHeader is the M3 deprecated header
+	// DeprecatedHeader is the M3 deprecated header.
 	DeprecatedHeader = "M3-Deprecated"
 
+	// LimitMaxTimeseriesHeader is the M3 limit timeseries header that limits
+	// the number of time series returned by each storage node.
+	LimitMaxTimeseriesHeader = "M3-Limit-Max-Timeseries"
```
Comment (richardartoul) [inline src/query/api/v1/handler/headers.go]: I think `M3-Limit-Max-Series` or just `M3-Max-Series` to be easier to remember

File: src/query/api/v1/handler/prometheus/native/list_tags.go [hunk 6]
```diff
@@ -48,13 +48,15 @@ var (
 
 // ListTagsHandler represents a handler for list tags endpoint.
 type ListTagsHandler struct {
-	storage storage.Storage
-	nowFn   clock.NowFn
+	storage             storage.Storage
```
Comment (richardartoul) [inline src/query/api/v1/handler/prometheus/native/list_tags.go]: This new limit wont affect aggregate endpoints will it?

File: src/query/api/v1/handler/search.go [hunk 7]
```diff
@@ -46,24 +46,32 @@ const (
...
 func (h *SearchHandler) ServeHTTP(w http.ResponseWriter, r *http.Request) {
 	logger := logging.WithContext(r.Context())
 
-	query, rErr := h.parseBody(r)
-	if rErr != nil {
-		logger.Error("unable to parse request", zap.Error(rErr))
+	query, parseBodyErr := h.parseBody(r)
+	opts, parseURLParamsErr := h.parseURLParams(r)
+	// NB(r): Use a loop here to avoid two err handling code paths
+	for _, rErr := range []*xhttp.ParseError{parseBodyErr, parseURLParamsErr} {
```
Comment (richardartoul) [inline src/query/api/v1/handler/search.go]: Took me longer to digest this than two error handlings paths, and it possibly allocates? Would probably just do the normal thing here personally or at least do

```golang
for _, rErr := range [2]*xhttp.ParseError{parseBodyErr, parseURLParamsErr} {
...
}
```

to be sure it does not allocate

File: src/query/api/v1/handler/search.go [hunk 8]
```diff
@@ -90,24 +98,19 @@ func (h *SearchHandler) parseBody(r *http.Request) (*storage.FetchQuery, *xhttp.
...
-		if err != nil {
-			limit = defaultLimit
+func (h *SearchHandler) parseURLParams(r *http.Request) (*storage.FetchOptions, *xhttp.ParseError) {
+	fetchOpts, err := h.fetchOptionsBuilder.NewFetchOptions(r)
+	if err != nil {
+		return nil, err
+	}
+
+	if str := r.URL.Query().Get("limit"); str != "" {
+		if limit, err := strconv.Atoi(str); err == nil {
```
Comment (richardartoul) [inline src/query/api/v1/handler/search.go]: honestly if they pass the correct query URL name but we can't parse it into a number, it should just return an error

File: src/query/models/query_context.go [hunk 9]
```diff
@@ -35,24 +35,37 @@ type QueryContext struct {
 	Ctx      context.Context
 	Scope    tally.Scope
 	Enforcer cost.ChainedEnforcer
+	Options  QueryContextOptions
+}
+
+// QueryContextOptions is a set of optionally set options for the query
```
Comment (richardartoul) [inline src/query/models/query_context.go]: `QueryContextOptions contains optional configuration for the query context`

File: src/query/server/server.go [hunk 10]
```diff
@@ -727,25 +736,40 @@ func remoteClient(
...
 ) (*grpc.Server, error) {
 	logger.Info("creating gRPC server")
-	server := tsdbRemote.CreateNewGrpcServer(storage, poolWrapper)
-	waitForStart := make(chan struct{})
-	var startErr error
+	server := tsdbRemote.CreateNewGrpcServer(storage,
+		queryContextOptions, poolWrapper)
+
+	var (
+		waitForStart = make(chan struct{}, 1)
```
Comment (richardartoul) [inline src/query/server/server.go]: Can you just make this a `chan error` and get rid of the startErr atomic stuff? Seems confusing for no reason.

```golang
// StartNewGrpcServer starts server on given address, then notifies channel
func StartNewGrpcServer(
	server *grpc.Server,
	address string,
	waitForStart chan<- error{},
) error {
	lis, err := net.Listen("tcp", address)
	if err != nil {
                 waitForStart <-err
	}

	waitForStart <- nil
	return server.Serve(lis)
}
```

File: src/query/server/server.go [hunk 11]
```diff
@@ -727,25 +736,40 @@ func remoteClient(
...
+		startErr     atomic.Value
+		wg           sync.WaitGroup
+	)
+	startErr.Store(nil)
+	wg.Add(1)
```
Comment (richardartoul) [inline src/query/server/server.go]: I dont think you actually use this

File: src/query/api/v1/handler/search.go (hunk 7 above)
Comment (robskillington) [reply src/query/api/v1/handler/search.go]: Sure, I might actually just do a helper function like I have in `xerrors.FirstError` (only reason I didn't do that is that it expects `func FirstError(err ...error) error` but these are typed errors....

File: src/query/server/server.go (hunk 10 above)
Comment (robskillington) [reply src/query/server/server.go]: The whole thing is kinda messed up, I wanted to introduce least amount of changes possible.

I'll just break it up into two steps and make the grpc server take the listener.

File: src/query/api/v1/handler/search.go (hunk 8 above)
Comment (robskillington) [reply src/query/api/v1/handler/search.go]: Yeah agreed, will change what this behavior used to do to match that.

File: src/query/server/server.go (hunk 11 above)
Comment (robskillington) [reply src/query/server/server.go]: Ta, I'm going to rip all this stuff out thankfully - good catch though.

File: src/query/api/v1/handler/prometheus/native/list_tags.go (hunk 6 above)
Comment (robskillington) [reply src/query/api/v1/handler/prometheus/native/list_tags.go]: It will too yeah.

Comment (robskillington) [reply src/query/api/v1/handler/prometheus/native/list_tags.go]: But will correctly respect the size of the aggregate values, not the amount of documents scanned.

File: src/query/api/v1/handler/fetch_options.go (hunk 4 above)
Comment (robskillington) [reply src/query/api/v1/handler/fetch_options.go]: It does make more sense to be a URL parameter but I wanted this to be a part of any endpoint.  So I didn't want to necessarily have one endpoint using "limit" for something other than what I'm specifying then have them collide.

We can always go add a URL param later if we think we can avoid collisions, this was just a quick way to add it for now.

Comment (robskillington) [reply src/query/api/v1/handler/fetch_options.go]: Yeah, let me look around for it.  @benraskin92 do you know of any?

**Pull request merged**